Numeric literal expressions and literal suffixes (#1177)
rust-lang/reference: src/tokens.md

> **<sup>Lexer</sup>**\
> FLOAT_LITERAL :\
> &nbsp;&nbsp; &nbsp;&nbsp; DEC_LITERAL `.`

> >   _(not immediately followed by `.`, `_` or an [identifier] or [keyword][keywords]_)\

Reviewer: To be pedantic here, I think it is an identifier, raw identifier, or keyword.  We don't have have anything to succinctly refer to that. What do you think about adding a new rule to the identifiers chapter?  Maybe something like:

> IDENTIFIER_OR_KEYWORD_OR_RAW: IDENTIFIER_OR_KEYWORD | RAW_IDENTIFIER

I...can't think of a good name for that, maybe you can improve that.  That rule can then be referenced here, and the macros-by-example chapter (where it refers to `ident` in the Metavariables section).

Author: Ah, it could be a byte literal or a raw string literal too.

I think it will be best to switch to just talking about the following character directly, rather than how the following input would be tokenised (so matching the implementation more closely).

A _floating-point literal_ has one of three forms:

* A _decimal literal_ followed by a period character `U+002E` (`.`). This is
  optionally followed by another decimal literal, with an optional _exponent_.
* A single _decimal literal_ followed by an _exponent_.
* A single _decimal literal_ (in which case a suffix is required).

Like integer literals, a floating-point literal may be followed by a
suffix, so long as the pre-suffix part does not end with `U+002E` (`.`).
There are two valid _floating-point suffixes_: `f32` and `f64` (the names of the 32-bit and 64-bit [primitive floating-point types][floating-point types]).
See [literal expressions] for the effect of these suffixes.

Examples of floating-point literals of various forms:

```rust
123.0f64;
0.1f64;
0.1f32;
12E+99_f64;
5f32;
let x: f64 = 2.;
```

This last example is different because it is not possible to use the suffix
syntax with a floating point literal ending in a period. `2.f64` would attempt
to call a method named `f64` on `2`.

Note that `-1.0`, for example, is analyzed as two tokens: `-` followed by `1.0`.

#### Number pseudoliterals

> **<sup>Lexer</sup>**\
> NUMBER_PSEUDOLITERAL :\

> > &nbsp;&nbsp; &nbsp;&nbsp; DEC_LITERAL ( . DEC_LITERAL)? FLOAT_EXPONENT NUMBER_PSEUDOLITERAL_SUFFIX\

Reviewer: ```suggestion
> &nbsp;&nbsp; &nbsp;&nbsp; DEC_LITERAL ( . DEC_LITERAL)<sup>?</sup> FLOAT_EXPONENT NUMBER_PSEUDOLITERAL_SUFFIX\
```

> > &nbsp;&nbsp; | DEC_LITERAL ( . DEC_LITERAL)? NUMBER_PSEUDOLITERAL_SUFFIX_NO_E\

Reviewer: ```suggestion
> &nbsp;&nbsp; | DEC_LITERAL ( . DEC_LITERAL)<sup>?</sup> NUMBER_PSEUDOLITERAL_SUFFIX_NO_E\
```

> &nbsp;&nbsp; | ( BIN_LITERAL | OCT_LITERAL | HEX_LITERAL ) NUMBER_PSEUDOLITERAL_SUFFIX_NO_E
>
> NUMBER_PSEUDOLITERAL_SUFFIX :\

> > &nbsp;&nbsp; IDENTIFIER_OR_KEYWORD <sub>_not matching INTEGER_SUFFIX or FLOAT_SUFFIX_</sub>

Reviewer: Unless I'm misunderstanding this, it seems to not cover something like `1.3E10u64`, which seems to be a single token.

Author: You're right. It also wasn't covering non-decimal with a float-type suffix, eg `0b1111_f32`.

I think the best way to write it will be to add back the extra suffixes rather than inventing more kinds of `NUMBER_PSEUDOLITERAL_SUFFIX_NO_XXX`, something like this:

````
NUMBER_PSEUDOLITERAL :
     DEC_LITERAL ( . DEC_LITERAL)? FLOAT_EXPONENT ( NUMBER_PSEUDOLITERAL_SUFFIX | INTEGER_SUFFIX )
   | DEC_LITERAL . DEC_LITERAL ( NUMBER_PSEUDOLITERAL_SUFFIX_NO_E | INTEGER SUFFIX )
   | DEC_LITERAL NUMBER_PSEUDOLITERAL_SUFFIX_NO_E
   | ( BIN_LITERAL | OCT_LITERAL | HEX_LITERAL ) ( NUMBER_PSEUDOLITERAL_SUFFIX_NO_E | FLOAT_SUFFIX )
````

>
> NUMBER_PSEUDOLITERAL_SUFFIX_NO_E :\

> > &nbsp;&nbsp; NUMBER_PSEUDOLITERAL_SUFFIX <sub>_not beginning with `e` or `E`_</sub>

Reviewer: This "beginning with e" is a bit surprising to me.  So strange!

> As described [above](#suffixes), tokens with the same form as numeric literals other than in the content of their suffix are accepted by the lexer (with the exception of some reserved forms described below).

Reviewer: Perhaps it would be good to make it clear that these aren't used for anything other than tokenizing?  I might perhaps write something like this:

```suggestion
Tokenization of numeric literals allows arbitrary suffixes as described in the grammar above.
These values generate valid tokens, but are not a valid [literal expression], so are usually an error except as macro arguments.
```

Examples of such tokens:
```rust,compile_fail
0invalidSuffix;
123AFB43;
0b010a;
0xAB_CD_EF_GH;
2.0f80;
2e5f80;
2e5e6;
2.0e5e6;
```

#### Reserved forms similar to number literals

> **<sup>Lexer</sup>**\
> RESERVED_NUMBER :\
> &nbsp;&nbsp; &nbsp;&nbsp; BIN_LITERAL \[`2`-`9`&ZeroWidthSpace;]\
> &nbsp;&nbsp; | OCT_LITERAL \[`8`-`9`&ZeroWidthSpace;]\
> &nbsp;&nbsp; | ( BIN_LITERAL | OCT_LITERAL | HEX_LITERAL ) `.` \
> &nbsp;&nbsp; &nbsp;&nbsp; &nbsp;&nbsp; _(not immediately followed by `.`, `_` or an [identifier] or [keyword][keywords]_)\
> &nbsp;&nbsp; | ( BIN_LITERAL | OCT_LITERAL ) `e`\
> &nbsp;&nbsp; | `0b` `_`<sup>\*</sup> _end of input or not BIN_DIGIT_\
> &nbsp;&nbsp; | `0o` `_`<sup>\*</sup> _end of input or not OCT_DIGIT_\
> &nbsp;&nbsp; | `0x` `_`<sup>\*</sup> _end of input or not HEX_DIGIT_\
> &nbsp;&nbsp; | DEC_LITERAL ( . DEC_LITERAL)? (`e`|`E`) (`+`|`-`)? _end of input or not DEC_DIGIT_

> The following lexical forms similar to number literals are _reserved forms_:

Reviewer: This introduction a little unclear to me.  Perhaps it could explain what these are up-front?  Maybe something like:

```suggestion
The following lexical forms similar to number literals are _reserved forms_.
Due to the possible ambiguity these raise, they are rejected by the tokenizer instead of being interpreted as separate tokens.
```
